python/ray/tune/tests/test_tune_save_restore.py

```diff
@@ -11,6 +11,123 @@
+            self.MockTrainable,
+            num_samples=num_samples,
+            restore=restore,
+            name=exp_name,
+            stop={
+                "training_iteration": 1
+            },
+            checkpoint_freq=1,
+            local_dir=local_dir,
+            config={
+                "env": "CartPole-v0",
+                "log_level": "DEBUG"
+            }).trials
+
+        exp_dir = os.path.join(absolute_local_dir, exp_name)
+
+        checkpoint_paths = []
+
+        for trial in trials:
+            self.assertIsNone(trial.error_file)
+            self.assertEqual(trial.local_dir, exp_dir)
+            checkpoint_path = os.path.join(trial.logdir,
+                                           "checkpoint_1/checkpoint-1")
+            self.assertTrue(os.path.isfile(checkpoint_path))
+            checkpoint_path.append(checkpoint_path)
+
+        return trials, checkpoint_paths
+
+    def testSimpleRestore(self):
+        # test base functionality of restore=<path_to_checkpoint>
+        exp_name = self.prefix + "SimpleRestore"
+        local_dir = "./test_simple_restore"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self.assertFalse(os.path.exists(absolute_local_dir))
+        trials, paths_to_checkpoints = self._train(
+            exp_name,
+            local_dir=local_dir,
+            absolute_local_dir=absolute_local_dir)
+        new_trials, new_paths_to_checkpoints = self._train(
+            exp_name,
+            local_dir=local_dir,
+            absolute_local_dir=absolute_local_dir,
+            restore=paths_to_checkpoints[0])
+        # ensure the restored trial was given a new ID
+        self.assertNotEqual(trials[0].trial_id, new_trials[0].trial_id)
+        # the new trial's name should have restore=<reference_to_restored_trial>
+        self.assertTrue('restore=' in new_paths_to_checkpoints[0])
+        self.assertTrue(trials[0].trial_id in new_paths_to_checkpoints[0])
+
+    def testWarmStartGridSearch(self):
+        exp_name = self.prefix + "WarmStartGridSearch"
+        local_dir = "./warm_start_grid_search"
+        absolute_local_dir = os.path.abspath(local_dir)
+        self.assertFalse(os.path.exists(absolute_local_dir))
+        # train a small number of trials
+        trials, paths_to_checkpoints = self._train(
+            exp_name,
+            local_dir=local_dir,
+            absolute_local_dir=absolute_local_dir,
+            num_samples=4)
+        # start a new set of trials using the checkpoints
+        # as a starting point
+        new_trials, new_checkpoints = self._train(
+            exp_name, restore=tune.grid_search(paths_to_checkpoints))
+        for old_trial, old_checkpoint in zip(trials, paths_to_checkpoints):
+            used = False
+            for new_trial, new_checkpoint in zip(new_trials, new_checkpoints):
+                self.assertNotEqual(old_trial.trial_id, new_trial.trial_id)
+                self.assertTrue('restore=' in new_checkpoint)
```

Comment:
be sure to run `ci/scripts/format.sh` - we only use single quotes in the codebase.

Reply:
```
~/Repositories/ray$ ./scripts/format.sh 
WARNING: Ray uses clang-format 7.0.0, You currently are using 9.0.0. This might generate different results.
From github.com:ray-project/ray
 * branch              master     -> FETCH_HEAD
Usage: flake8 [options] file file ...

flake8: error: no such option: --inline-quotes
~/Repositories/ray$ flake8 --version
3.7.7 (mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.7.3 on Darwin
```

*scratches head*

Reply:
Ah... `pip install flake8-quotes`?

Reply:
Nope - still the same output.

Reply:
Are you sure flake8 quotes is installed? For me, this is what shows up.
```
3.7.7 (flake8_quotes: 2.0.1, mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.7.3 on Darwin
```